apiserver/facades/agent/upgrader/watchapi_test.go

```diff
@@ -0,0 +1,228 @@
+		}}
+	results, err := s.api().WatchAPIVersion(context.Background(), args)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(results, gc.DeepEquals, params.NotifyWatchResults{
+		Results: []params.NotifyWatchResult{
+			{NotifyWatcherId: "42"},
+		},
+	})
+}
+
+func (s *upgraderWatchSuite) TestWatchAPIVersionUnit(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	done := make(chan struct{})
+	defer close(done)
+	ch := make(chan struct{})
+	w := watchertest.NewMockNotifyWatcher(ch)
+
+	tag := names.NewUnitTag("test/1")
+
+	s.agentService.EXPECT().WatchUnitTargetAgentVersion(gomock.Any(), tag.Id()).DoAndReturn(func(_ context.Context, _ string) (watcher.Watcher[struct{}], error) {
+		time.AfterFunc(coretesting.ShortWait, func() {
+			// Send initial event.
+			select {
+			case ch <- struct{}{}:
+			case <-done:
+				c.ExpectFailure("watcher (unit) did not fire")
+			}
+		})
+		return w, nil
+	})
+	s.watcherRegistry.EXPECT().Register(gomock.Any()).Return("4", nil)
+
+	args := params.Entities{
+		Entities: []params.Entity{
+			{Tag: tag.String()},
+		}}
+	results, err := s.api().WatchAPIVersion(context.Background(), args)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(results, gc.DeepEquals, params.NotifyWatchResults{
+		Results: []params.NotifyWatchResult{
+			{NotifyWatcherId: "4"},
+		},
+	})
+}
+
+func (s *upgraderWatchSuite) TestWatchAPIVersionControllerModelAgent(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	done := make(chan struct{})
+	defer close(done)
+	chC := make(chan struct{})
+	chM := make(chan struct{})
+	wc := watchertest.NewMockNotifyWatcher(chC)
+	wm := watchertest.NewMockNotifyWatcher(chM)
+
+	s.agentService.EXPECT().WatchModelTargetAgentVersion(gomock.Any()).DoAndReturn(func(_ context.Context) (watcher.Watcher[struct{}], error) {
+		time.AfterFunc(coretesting.ShortWait, func() {
+			// Send initial event.
+			select {
+			case chC <- struct{}{}:
+			case <-done:
+				c.ExpectFailure("watcher (controller) did not fire")
+			}
+		})
+		return wc, nil
+	})
+	s.agentService.EXPECT().WatchModelTargetAgentVersion(gomock.Any()).DoAndReturn(func(_ context.Context) (watcher.Watcher[struct{}], error) {
+		time.AfterFunc(coretesting.ShortWait, func() {
+			// Send initial event.
+			select {
+			case chM <- struct{}{}:
+			case <-done:
+				c.ExpectFailure("watcher (model) did not fire")
+			}
+		})
+		return wm, nil
+	})
+	s.watcherRegistry.EXPECT().Register(gomock.Any()).Return("2", nil)
+	s.watcherRegistry.EXPECT().Register(gomock.Any()).Return("1", nil)
+
+	args := params.Entities{
+		Entities: []params.Entity{
+			{Tag: coretesting.ControllerTag.String()},
+			{Tag: coretesting.ModelTag.String()},
+		}}
+	results, err := s.api().WatchAPIVersion(context.Background(), args)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(results, gc.DeepEquals, params.NotifyWatchResults{
+		Results: []params.NotifyWatchResult{
+			{NotifyWatcherId: "2"},
+			{NotifyWatcherId: "1"},
+		},
+	})
+}
+
+func (s *upgraderWatchSuite) TestWatchAPIVersionTagInvalid(c *gc.C) {
```

Comment:
You test a fully invalid tag, but what about a type like a User that is a valid entity, but not valid as a API version to watch?

Reply:
added